src/couch_stats/src/couch_stats_util.erl

```diff
@@ -0,0 +1,190 @@
+% the License.
+
+-module(couch_stats_util).
+
+-export([
+    % Load metrics from apps
+    create_metrics/1,
+    load_metrics_for_applications/0,
+    metrics_changed/2,
+
+    % Get various metric types
+    get_counter/2,
+    get_gauge/2,
+    get_histogram/2,
+
+    % Get histogram interval config settings
+    histogram_interval_sec/0,
+    histogram_safety_buffer_size_sec/0,
+    reset_histogram_interval_sec/0,
+
+    % Manage the main stats (metrics) persistent term map
+    replace_stats/1,
+    stats/0,
+    histograms/1,
+
+    % Fetch stats values
+    fetch/3,
+    sample/4
+]).
+
+-define(DEFAULT_INTERVAL_SEC, 10).
+
+% Histogram types
+-define(HIST, histogram).
+-define(CNTR, counter).
+-define(GAUGE, gauge).
+
+% Safety buffer before and after current window to prevent
+% overwrites from the cleaner process
+-define(HIST_WRAP_BUFFER_SIZE_SEC, 5).
+
+% Persistent term keys
+-define(STATS_KEY, {?MODULE, stats}).
+-define(HIST_TIME_INTERVAL_KEY, {?MODULE, hist_time_interval}).
+
+load_metrics_for_applications() ->
+    Apps = [element(1, A) || A <- application:loaded_applications()],
+    lists:foldl(fun load_metrics_for_application_fold/2, #{}, Apps).
+
+load_metrics_for_application_fold(AppName, #{} = Acc) ->
+    case code:priv_dir(AppName) of
+        {error, _Error} ->
+            Acc;
+        Dir ->
+            case file:consult(Dir ++ "/stats_descriptions.cfg") of
+                {ok, Descriptions} ->
+                    DescMap = maps:map(
+                        fun(_, TypeDesc) ->
+                            Type = proplists:get_value(type, TypeDesc, counter),
+                            Desc = proplists:get_value(desc, TypeDesc, <<>>),
+                            {Type, Desc}
+                        end,
+                        maps:from_list(Descriptions)
+                    ),
+                    maps:merge(Acc, DescMap);
+                {error, _Error} ->
+                    Acc
+            end
+    end.
+
+metrics_changed(#{} = Map1, #{} = Map2) when map_size(Map1) =/= map_size(Map2) ->
+    % If their sizes are differently they are obvioulsy not the same
+    true;
+metrics_changed(#{} = Map1, #{} = Map2) when map_size(Map1) =:= map_size(Map2) ->
+    % If their intersection size is not the same as their individual size
+    % they are also not the same
+    map_size(maps:intersect(Map1, Map2)) =/= map_size(Map1).
+
+get_counter(Name, #{} = Stats) ->
+    get_metric(Name, ?CNTR, Stats).
+
+get_gauge(Name, #{} = Stats) ->
+    get_metric(Name, ?GAUGE, Stats).
+
+get_histogram(Name, #{} = Stats) ->
+    get_metric(Name, ?HIST, Stats).
+
+get_metric(Name, Type, Stats) when is_atom(Type), is_map(Stats) ->
+    case maps:get(Name, Stats, unknown_metric) of
+        {FoundType, Metric, _Desc} when FoundType =:= Type ->
+            {ok, Metric};
+        {OtherType, _, _} ->
+            error_logger:error_msg("invalid metric: ~p ~p =/= ~p", [Name, Type, OtherType]),
```

Comment:
I think it's because it creates a circular dependency between `couch_log` and `couch_stats`. I was surprised by `error_logger` as well but stuck with it since it that's what `couch_stats` used already.